pyro/distributions/gaussian_scale_mixture.py

```diff
@@ -0,0 +1,145 @@
+    That is, the covariance matrix of the k'th component is given by
+
+    Sigma_ii = (lambda_k * scale_i) ** 2
+
+    where `lambda`_k is a positive scale factor and `scale`_i are positive
+    scale parameters shared between all K components. The mixture weights are
+    controlled by a K-dimensional vector of softmax logits, `logits`. This
+    distribution implements pathwise derivatives for samples from the distribution.
+    This distribution does not currently support batched parameters.
+
+    See reference [1] for details on the implementations of the pathwise
+    derivative. Please consider citing this reference if you use the pathwise
+    derivative in your research.
+
+    [1] Pathwise Derivatives for Multivariate Distributions, Martin Jankowiak &
+    Theofanis Karaletsos. arXiv:1806.01856
+
+    Note that this distribution supports both even and odd dimensions, but the
+    former should be more a bit higher precision, since it doesn't use any erfs in
+    the backward call.
+
+    :param torch.tensor scale: D-dimensional vector of scale
+    :param torch.tensor logits: K-dimensional vector of logits
+    :param torch.tensor lambdas: K-dimensional matrix of sigma multipliers
+    """
+    has_rsample = True
+    arg_constraints = {"lambdas": constraints.positive, "scale": constraints.positive,
+                       "logits": constraints.real}
+
+    def __init__(self, scale, logits, lambdas):
+        self.dim = scale.size(0)
+        assert(scale.dim() == 1), "The scale parameter in GaussianScaleMixture should be D dimensional"
+        assert(lambdas.dim() == 1), "The lambdas parameter in GaussianScaleMixture should be K dimensional"
+        assert(logits.dim() == 1), "The logits parameter in GaussianScaleMixture should be K dimensional"
+        assert(logits.shape == lambdas.shape), \
+            "The logits and lambdas parameters in GaussianScaleMixture should be K dimensional"
+        self.scale = scale
+        self.logits = logits
+        self.lambdas = lambdas
+        self.coeffs = self._compute_coeffs()
+        self.categorical = Categorical(logits=logits)
+        super(GaussianScaleMixture, self).__init__()
+
+    def _compute_coeffs(self):
+        """
+        These coefficients are used internally in the backward call.
+        """
+        dimov2 = int(self.dim / 2)  # this is correct for both even and odd dimensions
+        coeffs = torch.ones(dimov2)
+        for k in range(dimov2 - 1):
+            coeffs[k + 1:] *= self.dim - 2 * (k + 1)
+        return coeffs
+
+    def log_prob(self, value):
+        assert value.dim() == 1 and value.size(0) == self.dim
+        epsilon_sqr = torch.pow(value / self.scale, 2.0).sum()
+        lambdas_power = torch.pow(self.lambdas, -self.dim)
+        result = lambdas_power * self.categorical.probs * \
+            torch.exp(-0.5 * epsilon_sqr / torch.pow(self.lambdas, 2.0))  # K
+        result = torch.log(result.sum())
```

Comment:
nit: I'm surprised you didn't use `log_sum_exp()` here

Reply:
left TODO to do this once torch.logsumexp is in pytorch release